### DISCUSSION TITLE
best way to test example request data

### DISCUSSION THREAD

**Discussion by havok2063** [fastapi/fastapi#9690] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Query


app = FastAPI()


@app.get("/items/")
async def get_item(item_id: int = Query(..., description='a number', example=10)):
    return {"item_id": item_id}
```

### Description

Is there way to create tests out of the [request example data ](https://fastapi.tiangolo.com/tutorial/schema-extra-example/) you've declared for the OpenAPI docs?  Or if not, what's the best way to do that?  I'd like to make sure my examples continue to work after code changes, but avoid manually duplicating the example data as tests, since they risk getting out of sync.  

### Operating System

macOS

### FastAPI Version

0.97.0

### Python Version

3.9

**Comment by ahmedabdou14** [comment]
If i understood your question correct, i got a workaround it but it might not be very convenient.

You can add the `openapi_url` argument to your app.

```python
    app = FastAPI(
        title=settings.PROJECT_NAME,
        description="Project description",
        openapi_url=f"{settings.API_V1_STR}/openapi.json",
        docs_url=f"{settings.API_V1_STR}/docs",
        redoc_url=f"{settings.API_V1_STR}/redoc",
    )
```

This will return all data about your endpoints and schemas you used. Just like `Swagger` but in `JSON` data.

Example from [FastApi doc:](https://fastapi.tiangolo.com/tutorial/first-steps/?h=openapi#check-the-openapijson)

```json
{
    "openapi": "3.0.2",
    "info": {
        "title": "FastAPI",
        "version": "0.1.0"
    },
    "paths": {
        "/items/": {
            "get": {
                "responses": {
                    "200": {
                        "description": "Successful Response",
                        "content": {
                            "application/json": {



...
```

Here is an example endpoint that takes in an item in the request body and returns its count:

```python
@router.get("/")
async def get_item_count(
    item: Annotated[
        dict,
        Body(
            example={
                "item_id": "123",
                "item_name": "itemname",
                "item_count": 10,
            }
        )
    ]
) -> int:
    return item["item_count"]
```

When you access `/openapi.json` endpoint, it will output all information about the server in details, here is the the data specific for the `get_item_count` endpoint:

```json
"/api/v1/items/": {
    "get": {
        "tags": [
            "item"
        ],
        "summary": "Get Item Count",
        "operationId": "get_item_count_api_v1_items__get",
        "requestBody": {
            "content": {
                "application/json": {
                    "schema": {
                        "title": "Item",
                        "type": "object"
                     },
                    "example": {
                        "item_id": "123",
                        "item_name": "itemname",
                        "item_count": 10
                     }
                }
              },
        "required": true
    ....
}
```

Finally, you can use this api and extract the data you need from them, _in this case the example object_,  and enforce specific values for each endpoint.
Honestly, I don't know if this effort is worth it, but it might be a work around.

**Comment by manvillej** [comment, marked as answer]
Here is my suggestion:

1. use the Pydantic schema_extra.
```
# models/Item.py
class Item(BaseModel):
    name: str
    description: str | None = None
    price: float
    tax: float | None = None

    class Config:
        schema_extra = {
            "example": {
                "name": "Foo",
                "description": "A very nice Item",
                "price": 35.4,
                "tax": 3.2,
            }
      }
```

2. Separate your example from your model file.
```
# models/item.py
ItemExample =  {
                "name": "Foo",
                "description": "A very nice Item",
                "price": 35.4,
                "tax": 3.2,
            }

class Item(BaseModel):
    name: str
    description: str | None = None
    price: float
    tax: float | None = None

    class Config:
        schema_extra = {
            "example": ItemExample
      }
```

3. import the example & test. (using pytest as an example)
```
# conftest.py
from app.models.item import ItemExample
@pytest.fixture()
def item_example():
    yield ItemExample

```

```
def test_create_item(test_app, item_example):
    request_payload = item_example
    response = test_app.post(
        "/items/",
        content=json.dumps(request_payload)
    )
    assert response.status_code == 201
    response_json = response.json()
    assert response_json['id']
```
